File: packages/shared/src/citations.ts
// Deterministic citation verifier.
//
// The spec is explicit: this is NOT an LLM call. For every citation in a
// draft, we look up the cited policy and check the quote appears verbatim
// (modulo minor whitespace/punctuation variance) in the policy body. If the
// quote can't be found, the citation is invalid and the draft is rejected
// for redrafting.
//
// This single deterministic check is the load-bearing defense against the
// worst hallucination risk (fabricated policy citations).

import type { Citation, CitationVerification, InvalidCitation } from "./types.js";

export interface PolicyDoc {
  id: string;
  body: string;
}

// Normalize whitespace and a small set of typographically-equivalent chars
// (curly quotes, dashes) so that LLM reformatting of a real quote doesn't
// cause false negatives. We do NOT lowercase, drop punctuation, or strip
// words — those would be too permissive.
export function normalizeForCitationMatch(s: string): string {
  return s
    .replace(/[‘’‚‛′]/g, "'") // curly singles
    .replace(/[“”„‟″]/g, '"') // curly doubles
    .replace(/[–—−]/g, "-") // en/em/minus dash
    .replace(/\s+/g, " ")
    .trim();
}

const MIN_QUOTE_CHARS = 20;

export function verifyCitations(
  citations: Citation[],
  policies: PolicyDoc[],
): CitationVerification {
  const policyById = new Map(policies.map((p) => [p.id, normalizeForCitationMatch(p.body)]));
  const invalid: InvalidCitation[] = [];
  let validCount = 0;

  for (const c of citations) {
    const policy = policyById.get(c.policyId);
    if (!policy) {
      invalid.push({ citation: c, reason: `policy ${c.policyId} not in retrieval set` });
      continue;
    }
    const q = normalizeForCitationMatch(c.quote);
    if (q.length < MIN_QUOTE_CHARS) {
      invalid.push({
        citation: c,
        reason: `quote too short (<${MIN_QUOTE_CHARS} chars) — not specific enough`,
      });
      continue;
    }
    if (!policy.includes(q)) {
      invalid.push({ citation: c, reason: "quote not found verbatim in cited policy" });
      continue;
    }
    validCount += 1;
  }

  return { allValid: invalid.length === 0, validCount, invalidCitations: invalid };
}

File: apps/worker/src/overturn_worker/citations.py
"""Deterministic citation verifier — Python mirror of `packages/shared/src/citations.ts`.

Both implementations follow the same normalization rules so a draft produced
on the worker matches a draft uploaded via web. Tests cross-check the two
behaviors on identical fixtures.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_MIN_QUOTE_CHARS = 20

_CURLY_SINGLES = "‘’‚‛′"
_CURLY_DOUBLES = "“”„‟″"
_DASHES = "–—−"


def normalize(s: str) -> str:
    s = re.sub(f"[{_CURLY_SINGLES}]", "'", s)
    s = re.sub(f"[{_CURLY_DOUBLES}]", '"', s)
    s = re.sub(f"[{_DASHES}]", "-", s)
    s = re.sub(r"\s+", " ", s).strip()
    return s


@dataclass
class Citation:
    policy_id: str
    quote: str
    source_url: str | None = None
    page: str | int | None = None


@dataclass
class InvalidCitation:
    citation: Citation
    reason: str


@dataclass
class Verification:
    all_valid: bool
    valid_count: int
    invalid_citations: list[InvalidCitation]


@dataclass
class PolicyDoc:
    id: str
    body: str


def verify_citations(citations: list[Citation], policies: list[PolicyDoc]) -> Verification:
    by_id = {p.id: normalize(p.body) for p in policies}
    invalid: list[InvalidCitation] = []
    valid_count = 0

    for c in citations:
        body = by_id.get(c.policy_id)
        if body is None:
            invalid.append(
                InvalidCitation(c, f"policy {c.policy_id} not in retrieval set")
            )
            continue
        q = normalize(c.quote)
        if len(q) < _MIN_QUOTE_CHARS:
            invalid.append(
                InvalidCitation(
                    c,
                    f"quote too short (<{_MIN_QUOTE_CHARS} chars) — not specific enough",
                )
            )
            continue
        if q not in body:
            invalid.append(InvalidCitation(c, "quote not found verbatim in cited policy"))
            continue
        valid_count += 1

    return Verification(
        all_valid=len(invalid) == 0,
        valid_count=valid_count,
        invalid_citations=invalid,
    )
